File: BFS.cpp
#include <iostream>
#include <vector>
#include <queue>
using namespace std;

class Graph
{
	private: 
		int v;
		vector<int> *adj;

	public: Graph(int v)
	{
		this->v=v;
		adj=new vector<int>[v+1];
	}

	public : void addEdge(int source,int destination)
	{
		if(source>v||destination>v)
		{
			return;
		}
		adj[source].push_back(destination);
	}
	public: vector<int> getEdge(int source)
	{
		return adj[source];
	}
};
	int main()
	{
		int vertices;
		cin>>vertices;
		Graph graph(vertices);
		int e=vertices-1;
		int count=1;
		while(count<=e)
		{
			int source,destination;
			cin>>source>>destination;
			graph.addEdge(source,destination);
			count++;
		}
		int level;
		cin>>level;
		queue<int>q;
		q.push(1);
		bool visited[vertices+1];
		visited[1]=true;
		int count1=1,temp=0;
		int num=0;
		while(!q.empty())
		{
			int z=q.front();
			q.pop();
			vector<int> v=graph.getEdge(z);
			count1++;
			if(count1==level)
					{
						num=v.size();
						break;
					}
			for(int i=0;i<v.size();i++)
			{
				if(!visited[v[i]])
				{
					visited[v[i]]=true;
					q.push(v[i]);
				}
			}
		}
		cout<<num<<endl;
		return 0;
	}

File: BFS.java
import java.util.*;
import java.lang.*;
import java.io.*;
import java.util.HashMap;

class BFS
{
	private static Map<Integer,LinkedList<Integer>> adjacencyList;

	public BFS(int n)
	{
		adjacencyList= new HashMap<Integer,LinkedList<Integer>>();
		for(int i=0;i<n;i++)
		{
			adjacencyList.put(i+1,new LinkedList<Integer>());
		}
	}

	public void addEdge(int source ,int destination)
	{
		if(source>adjacencyList.size()||destination>adjacencyList.size())
		{
			System.out.println("Invalid input");
			return;
		}
		List<Integer> slist=adjacencyList.get(source);
		slist.add(destination);
		List<Integer> dlist=adjacencyList.get(destination);
		dlist.add(source);
	}

	public static List<Integer> getEdge(int source)
	{
		if(source>adjacencyList.size())
		{
			return null;
		}
		return adjacencyList.get(source);
	}

	public static void Bfs(int start, int vertices)
	{
		boolean[] visited=new boolean[vertices+1];
		LinkedList<Integer>bfs=new LinkedList<Integer>();
			visited[start]=true;
			bfs.add(start);
			while(bfs.size()!=0)
			{
				int z=bfs.poll();
				System.out.print(z+" ");
				Iterator<Integer> i=adjacencyList.get(z).listIterator();
				while(i.hasNext())
				{
					int n= i.next();
					if(!visited[n])
					{
						visited[n]=true;
						bfs.add(n);
					}
				}
			}
	}
	public static void main(String[] args)
	{
		Scanner sc= new Scanner(System.in);
		int n=sc.nextInt();
		BFS graph= new BFS(n);
		int m=sc.nextInt();
		int count=0;
		while(count<m)
		{
			int source=sc.nextInt();
			int destination=sc.nextInt();
			graph.addEdge(source,destination);
			count++;
		}
		Bfs(1,n);
		System.out.println();
	}
}